File: apps/server/src/lib/queue.ts
/**
 * Background jobs.
 *
 * Anything that must not slow down a message send — push delivery, link unfurls,
 * thumbnails, digests — is enqueued after the transaction commits and handled by the
 * worker process.
 */

import { Queue, type JobsOptions } from 'bullmq';
import { config } from '../config.ts';

const connection = { url: config.REDIS_URL };

export const QUEUE_NAME = 'blob-jobs';

export interface JobPayloads {
  notify: { messageId: string };
  unfurl: { messageId: string };
  thumbnail: { attachmentId: string };
  'sweep-orphans': Record<string, never>;
}

export type JobName = keyof JobPayloads;

const defaultOptions: JobsOptions = {
  attempts: 3,
  backoff: { type: 'exponential', delay: 2_000 },
  removeOnComplete: 500,
  removeOnFail: 1_000,
};

export const jobs = new Queue<JobPayloads[JobName], void, JobName>(QUEUE_NAME, {
  connection,
  defaultJobOptions: defaultOptions,
});

/** Enqueue and swallow failures: a lost unfurl must never fail a message send. */
export function enqueue<N extends JobName>(
  name: N,
  data: JobPayloads[N],
  options?: JobsOptions,
): void {
  void jobs.add(name, data, options).catch((err) => {
    console.error({ err, job: name }, 'failed to enqueue job');
  });
}

export async function closeQueue(): Promise<void> {
  await jobs.close();
}

File: apps/api/src/blob_api/lib/queue.py
"""Background jobs.

Anything that must not slow down a message send — push delivery, link unfurls,
thumbnails, digests — is enqueued after the transaction commits and handled by the
worker process.

Enqueueing never raises into a request: a lost unfurl must not fail a message send.
"""

from __future__ import annotations

import logging
from typing import Any

from arq import create_pool
from arq.connections import ArqRedis, RedisSettings

from ..config import settings

log = logging.getLogger("blob.queue")

_pool: ArqRedis | None = None


def redis_settings() -> RedisSettings:
    return RedisSettings.from_dsn(settings.REDIS_URL)


async def get_pool() -> ArqRedis:
    global _pool
    if _pool is None:
        _pool = await create_pool(redis_settings())
    return _pool


async def enqueue(job: str, *args: Any) -> None:
    try:
        pool = await get_pool()
        await pool.enqueue_job(job, *args)
    except Exception:
        log.warning("could not enqueue %s", job, exc_info=True)


async def close_queue() -> None:
    global _pool
    if _pool is not None:
        await _pool.aclose()
        _pool = None
